client/my-sites/themes/helpers.js

```diff
@@ -148,20 +149,45 @@ export function interlaceThemes( wpComThemes, wpOrgThemes, searchTerm, isLastPag
 		  )
 		: [];
 
-	const matchingTheme = wpComThemes.find( isMatchingTheme );
+	const interlacedThemes = [];
+
+	// 1. Exact match.
+	const matchingTheme =
+		wpComThemes.find( isMatchingTheme ) || validWpOrgThemes.find( isMatchingTheme );
+	if ( matchingTheme ) {
+		interlacedThemes.push( matchingTheme );
+	}
+
+	// 2. WP.com themes.
 	const restWpComThemes = matchingTheme
 		? wpComThemes.filter( ( theme ) => theme.id !== matchingTheme.id )
 		: wpComThemes;
-	const matchingWpOrgTheme = validWpOrgThemes.find( isMatchingTheme );
-	const restWpOrgThemes = matchingWpOrgTheme
-		? validWpOrgThemes.filter( ( theme ) => theme.id !== matchingWpOrgTheme.id )
-		: validWpOrgThemes;
-
-	return [
-		...( matchingTheme ? [ matchingTheme ] : [] ),
-		...( matchingWpOrgTheme ? [ matchingWpOrgTheme ] : [] ),
-		...restWpComThemes,
-		// Include WP.org themes after the last page of WP.com themes.
-		...( isEnabled( 'themes/interlaced-dotorg-themes' ) && isLastPage ? restWpOrgThemes : [] ),
-	];
+	interlacedThemes.push( ...restWpComThemes );
+
+	// 3. WP.org themes (only if the list of WP.com themes has reached the last page).
+	if ( isEnabled( 'themes/interlaced-dotorg-themes' ) && isLastPage ) {
+		// 3.1. WP.org block themes.
+		const restWpOrgBlockThemes = validWpOrgThemes.filter(
+			( theme ) =>
+				theme.id !== matchingTheme?.id &&
+				theme.taxonomies?.theme_feature?.some(
+					( feature ) => feature?.slug === 'full-site-editing'
+				)
+		);
+		interlacedThemes.push( ...restWpOrgBlockThemes );
+
+		// 3.2 WP.org classic themes.
+		const restWpOrgClassicThemes = validWpOrgThemes.filter(
+			( theme ) =>
+				theme.id !== matchingTheme?.id &&
+				( ! theme.taxonomies ||
+					! theme.taxonomies.theme_feature ||
+					theme.taxonomies.theme_feature.every(
+						( feature ) => feature?.slug !== 'full-site-editing'
+					) )
+		);
```

Comment:
Wonder if we could do a single loop here with foreach (or reduce even). 🤔 

E.g.

```js
const restWpOrgBlockThemes = [];
const restWpOrgClassicThemes = [];

wpOrgThemes.forEach( theme, => {
	if ( theme.id === matchingTheme?.id ) {
		return;
	}
	if ( theme.taxonomies?.theme_feature?.some(
		( feature ) => feature?.slug === 'full-site-editing'
	) {
		restWpOrgBlockThemes.push( theme );
	} else {
		restWpOrgClassicThemes.push( theme );
	}
} );
```

Reply:
https://github.com/Automattic/wp-calypso/pull/78263/commits/c9fe8395d3156e73db5c413c22057ba89123565f